src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs

```diff
@@ -1936,28 +1934,30 @@ internal static unsafe bool TryParseSingle(ReadOnlySpan<char> value, NumberStyle
             return true;
         }
 
-        internal static unsafe bool TryStringToNumber(ReadOnlySpan<char> value, NumberStyles styles, ref NumberBuffer number, NumberFormatInfo info)
+        internal static bool TryStringToNumber(ReadOnlySpan<char> value, NumberStyles styles, ref NumberBuffer number, NumberFormatInfo info)
         {
             Debug.Assert(info != null);
-            fixed (char* stringPointer = &MemoryMarshal.GetReference(value))
+
+            ReadOnlySpan<char> p = value;
+            if (!TryParseNumber(ref p, styles, ref number, info) || !TrailingZeros(p))
             {
-                char* p = stringPointer;
-                if (!TryParseNumber(ref p, p + value.Length, styles, ref number, info)
-                    || ((int)(p - stringPointer) < value.Length && !TrailingZeros(value, (int)(p - stringPointer))))
-                {
-                    number.CheckConsistency();
-                    return false;
-                }
+                number.CheckConsistency();
+                return false;
             }
 
             number.CheckConsistency();
             return true;
         }
 
         private static bool TrailingZeros(ReadOnlySpan<char> value, int index)
+        {
+            return TrailingZeros(value.Slice(index));
+        }
+
+        private static bool TrailingZeros(ReadOnlySpan<char> value)
         {
             // For compatibility, we need to allow trailing zeros at the end of a number string
-            for (int i = index; (uint)i < (uint)value.Length; i++)
+            for (int i = 0; (uint)i < (uint)value.Length; i++)
```

Comment:
I took a look in .NET codebase and I remember to read something about (uint) cast to eliminate bound check inside the for loop.
Is this still needed? Or it has been fixed in JIT?
Thanks

Reply:
It seems it is still needed